File: app/src/main/java/net/mmho/photomap2/PhotoViewActivity.java
package net.mmho.photomap2;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.provider.MediaStore;
import android.support.v4.view.MenuItemCompat;
import android.support.v4.view.ViewPager;
import android.support.v4.view.WindowCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.app.ActionBarActivity;
import android.support.v7.widget.ShareActionProvider;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

public class PhotoViewActivity extends ActionBarActivity{

    public static final String EXTRA_GROUP = "photo_group";
    public static final String EXTRA_POSITION = "position";

    private PhotoViewAdapter adapter;
    private ViewPager pager;
    private ShareActionProvider shareActionProvider;
    private PhotoGroup group;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle bundle = getIntent().getExtras();

        if(bundle==null){
            finish();
            return;
        }

        supportRequestWindowFeature(WindowCompat.FEATURE_ACTION_BAR_OVERLAY);
        setContentView(R.layout.fragment_photo_view);
        setSupportActionBar((Toolbar)findViewById(R.id.toolbar));
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        showActionBar(true);
        ActionBar bar = getSupportActionBar();
        if(bar!=null) bar.addOnMenuVisibilityListener(menuVisibilityListener);

        group = bundle.getParcelable(EXTRA_GROUP);
        if(group.address!=null){
            setTitle(AddressUtil.getTitle(group.address,this));
        }
        int position = bundle.getInt(EXTRA_POSITION);
        adapter = new PhotoViewAdapter(getSupportFragmentManager(), group);

        pager = (ViewPager) findViewById(R.id.photo_pager);
        pager.setAdapter(adapter);
        pager.setOnPageChangeListener(onPageChangeListener);
        pager.setCurrentItem(position);
        pager.setPageMargin(30);
        pager.setOnClickListener(onClickListener);
    }

    final private Handler handler = new Handler();
    final private Runnable runnable = new Runnable() {
        @Override
        public void run() {
            hideActionBar();
        }
    };

    private void showActionBar(boolean hide) {
        ActionBar bar = getSupportActionBar();
        if(bar!=null) bar.show();
        if(!hide) handler.removeCallbacks(runnable);
        else hideActionBarDelayed();
    }

    private void hideActionBar(){
        ActionBar bar = getSupportActionBar();
        if(bar!=null){
            bar.hide();
        }
    }

    private void hideActionBarDelayed(){
        final long DELAY=3*1000;
        handler.removeCallbacks(runnable);
        handler.postDelayed(runnable,DELAY);
    }


    final private ActionBar.OnMenuVisibilityListener menuVisibilityListener =
            new ActionBar.OnMenuVisibilityListener() {
                @Override
                public void onMenuVisibilityChanged(boolean isVisible) {
                    if(isVisible) showActionBar(false);
                    else hideActionBarDelayed();
                }
            };

    final private View.OnClickListener onClickListener =
            new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    ActionBar bar = getSupportActionBar();
                    if(bar!=null){
                        if(bar.isShowing()) hideActionBar();
                        else showActionBar(true);
                    }
                }
            };

    final private ViewPager.OnPageChangeListener onPageChangeListener =
            new ViewPager.OnPageChangeListener() {
                @Override
                public void onPageScrolled(int i, float v, int i2) {

                }

                @Override
                public void onPageSelected(int i) {
                    if(shareActionProvider!=null)
                        shareActionProvider.setShareIntent(setShareIntent(new Intent(),i));
                }

                @Override
                public void onPageScrollStateChanged(int i) {

                }
            };

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.photo_view_menu,menu);
        MenuItem share = menu.findItem(R.id.share);
        shareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(share);
        shareActionProvider.setShareIntent(setShareIntent(new Intent(),pager.getCurrentItem()));
        shareActionProvider.setSubUiVisibilityListener(new ShareActionProvider.SubUiVisibilityListener() {
            @Override
            public void onSubUiVisibilityChanged(boolean visible) {
                if(visible) showActionBar(false);
                else hideActionBarDelayed();
            }
        });

        return super.onCreateOptionsMenu(menu);
    }

    private Intent setShareIntent(Intent intent,int index){
        final long image_id = adapter.getItemID(index);
        final Uri uri = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                String.valueOf(image_id));
        intent.setAction(Intent.ACTION_SEND);
        intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        intent.putExtra(Intent.EXTRA_STREAM,uri);
        intent.setType("image/jpeg");
        return intent;
    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Intent intent;
        switch (item.getItemId()){
        case R.id.map:
            intent = new Intent(this,PhotoMapActivity.class);
            setShareIntent(intent, pager.getCurrentItem());
            startActivity(intent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onBackPressed() {
        Intent i = new Intent();
        Bundle b = new Bundle();
        b.putInt(EXTRA_POSITION,pager.getCurrentItem());
        i.putExtras(b);
        setResult(RESULT_OK, i);
        super.onBackPressed();
    }
}

File: app/src/main/kotlin/net/mmho/photomap2/PhotoViewActivity.kt
package net.mmho.photomap2

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.provider.MediaStore
import androidx.core.view.MenuItemCompat
import androidx.viewpager.widget.ViewPager
import androidx.core.view.WindowCompat
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.ShareActionProvider
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.fragment_photo_view.*

class PhotoViewActivity : AppCompatActivity() {

    private var adapter: PhotoViewAdapter? = null
    private var shareActionProvider: ShareActionProvider? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val bundle = intent.extras

        if (bundle == null) {
            finish()
            return
        }

        supportRequestWindowFeature(WindowCompat.FEATURE_ACTION_BAR_OVERLAY)
        setContentView(R.layout.fragment_photo_view)
        setSupportActionBar(toolbar)
        showActionBar(true)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.addOnMenuVisibilityListener(menuVisibilityListener)

        val group = bundle.getParcelable<PhotoGroup>(EXTRA_GROUP)
        title = group?.title

        val position = bundle.getInt(EXTRA_POSITION)
        adapter = PhotoViewAdapter(supportFragmentManager, group)

        photo_pager.adapter = adapter
        photo_pager.addOnPageChangeListener(onPageChangeListener)
        photo_pager.currentItem = position
        photo_pager.pageMargin = 30
        photo_pager.setOnClickListener(onClickListener)
    }

    override fun onDestroy() {
        super.onDestroy()
        photo_pager.removeOnPageChangeListener(onPageChangeListener)
    }


    private val handler = Handler()
    private val runnable = Runnable { this.hideActionBar() }

    private fun showActionBar(hide: Boolean) {
        supportActionBar?.show()
        if (!hide)
            handler.removeCallbacks(runnable)
        else
            hideActionBarDelayed()
    }

    private fun hideActionBar() {
        supportActionBar?.hide()
    }

    private fun hideActionBarDelayed() {
        val delay = (3 * 1000).toLong()
        handler.removeCallbacks(runnable)
        handler.postDelayed(runnable, delay)
    }


    private val menuVisibilityListener = ActionBar.OnMenuVisibilityListener {
        when {
            it -> showActionBar(false)
            else -> hideActionBarDelayed()
        }
    }

    private val onClickListener = View.OnClickListener {
        val bar = supportActionBar
        when {
            bar == null -> {}
            bar.isShowing -> hideActionBar()
            else -> showActionBar(true)
        }
    }

    private val onPageChangeListener = object : ViewPager.SimpleOnPageChangeListener() {
        override fun onPageSelected(i: Int) {
            shareActionProvider?.setShareIntent(setShareIntent(Intent(), i))
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.photo_view_menu, menu)
        val share = menu.findItem(R.id.share)
        shareActionProvider = MenuItemCompat.getActionProvider(share) as ShareActionProvider
        shareActionProvider?.setShareIntent(setShareIntent(Intent(), photo_pager.currentItem))
        shareActionProvider?.setVisibilityListener { visible ->
            if (visible)
                showActionBar(false)
            else
                hideActionBarDelayed()
        }

        return super.onCreateOptionsMenu(menu)
    }

    private fun setShareIntent(intent: Intent, index: Int): Intent {
        val imageId = adapter?.getItemID(index)
        val uri = Uri.withAppendedPath(MediaStore.Images.Media.EXTERNAL_CONTENT_URI,
                imageId.toString())
        intent.action = Intent.ACTION_SEND
        intent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP)
        intent.putExtra(Intent.EXTRA_STREAM, uri)
        intent.type = "image/jpeg"
        return intent
    }


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val intent: Intent
        when (item.itemId) {
            R.id.map -> {
                intent = Intent(this, PhotoMapActivity::class.java)
                setShareIntent(intent, photo_pager.currentItem)
                startActivity(intent)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        val i = Intent()
        val b = Bundle()
        b.putInt(EXTRA_POSITION, photo_pager.currentItem)
        i.putExtras(b)
        setResult(Activity.RESULT_OK, i)
        super.onBackPressed()
    }

    companion object {
        const val EXTRA_GROUP = "photo_group"
        const val EXTRA_POSITION = "position"
    }
}
